test/integration/shoots/operations/clusterautoscaler.go

```diff
@@ -311,3 +310,146 @@ type reserveCapacityValues struct {
 	WorkerPool    string
 	TolerationKey string
 }
+
+// Resource is a collection of compute resource.
+type Resource struct {
+	MilliCPU int64
+	Memory   int64
+}
+
+// NewResource creates a Resource from ResourceList.
+func NewResource(rl corev1.ResourceList) *Resource {
+	r := &Resource{}
+	r.Add(rl)
+	return r
+}
+
+// Add adds ResourceList into Resource.
+func (r *Resource) Add(rl corev1.ResourceList) {
+	if r == nil {
+		return
+	}
+
+	for rName, rQuant := range rl {
+		switch rName {
+		case corev1.ResourceCPU:
+			r.MilliCPU += rQuant.MilliValue()
+		case corev1.ResourceMemory:
+			r.Memory += rQuant.Value()
+		default:
+			// do nothing, we track only cpu and memory
+		}
+	}
+}
+
+// SetMaxResource compares with ResourceList and takes max value for each Resource.
+func (r *Resource) SetMaxResource(rl corev1.ResourceList) {
+	if r == nil {
+		return
+	}
+
+	for rName, rQuantity := range rl {
+		switch rName {
+		case corev1.ResourceMemory:
+			if mem := rQuantity.Value(); mem > r.Memory {
+				r.Memory = mem
+			}
+		case corev1.ResourceCPU:
+			if cpu := rQuantity.MilliValue(); cpu > r.MilliCPU {
+				r.MilliCPU = cpu
+			}
+		default:
+			// do nothing, we track only cpu and memory
+		}
+	}
+}
+
+// NodeInfo is node level aggregated information.
+type NodeInfo struct {
+	Requested   *Resource
+	Allocatable *Resource
+}
+
+// Available returns the available resources on the Node.
+func (n *NodeInfo) Available() *Resource {
+	return &Resource{
+		MilliCPU: n.Allocatable.MilliCPU - n.Requested.MilliCPU,
+		Memory:   n.Allocatable.Memory - n.Requested.Memory,
+	}
+}
+
+func computeReserveCapacityResources(ctx context.Context, c kubernetes.Interface, nodes *corev1.NodeList) (*Resource, error) {
+	// Take half the Node allocatable capacity as requests for the deployment
+	nodeAllocatable := nodes.Items[0].Status.Allocatable
+	requestCPUMilli := nodeAllocatable.Cpu().MilliValue() / 2
+	requestMemory := nodeAllocatable.Memory().Value() / 2
+	podResource := &Resource{
+		MilliCPU: requestCPUMilli,
+		Memory:   requestMemory,
+	}
+
+	// Ensure that reserve-capacity Pod fits on the Nodes
+	for _, node := range nodes.Items {
+		nodeInfo, err := computeNodeInfo(ctx, c, &node)
+		if err != nil {
+			return nil, errors.Wrapf(err, "cannot to compute NodeInfo for Node '%s'", node.Name)
+		}
+
+		if !fitsPod(podResource, nodeInfo) {
+			availableResource := nodeInfo.Available()
+			if podResource.MilliCPU > availableResource.MilliCPU {
+				podResource.MilliCPU = availableResource.MilliCPU
```

Comment:
Can you give this some headroom?
```suggestion
				podResource.MilliCPU = availableResource.MilliCPU - 100
```